sdk/cosmos/azure-cosmos/azure/cosmos/_base.py

```diff
@@ -670,32 +670,35 @@ def ParsePaths(paths):
     return tokens
 
 
-def create_scope_from_url(url):
+def create_scope_from_url(url: str) -> str:
     parsed_url = urlsplit(url)
+    if not parsed_url.scheme or not parsed_url.hostname:
+        raise ValueError(f"Invalid URL scheme or hostname: {parsed_url!r}")
     return parsed_url.scheme + "://" + parsed_url.hostname + "/.default"
 
 
-def validate_cache_staleness_value(max_integrated_cache_staleness):
+def validate_cache_staleness_value(max_integrated_cache_staleness: Any) -> None:
     int(max_integrated_cache_staleness)  # Will throw error if data type cant be converted to int
     if max_integrated_cache_staleness < 0:
         raise ValueError("Parameter 'max_integrated_cache_staleness_in_ms' can only be an "
                          "integer greater than or equal to zero")
 
 
-def _stringify_auto_scale(offer: Dict[str, Any]) -> Any:
+def _stringify_auto_scale(offer: ThroughputProperties) -> str:
     auto_scale_params = None
     max_throughput = offer.auto_scale_max_throughput
     increment_percent = offer.auto_scale_increment_percent
     auto_scale_params = {"maxThroughput": max_throughput}
     if increment_percent is not None:
         auto_scale_params["autoUpgradePolicy"] = {"throughputPolicy": {"incrementPercent": increment_percent}}
     auto_scale_settings = json.dumps(auto_scale_params)
-
     return auto_scale_settings
 
 
-def _set_throughput_options(offer: Union[int, ThroughputProperties], request_options: Dict[str, Any]) -> Any:
-    if offer is not None:
+def _set_throughput_options(offer: Optional[Union[int, ThroughputProperties]], request_options: Dict[str, Any]) -> None:
```

Comment:
Because in the two places that this method is called, it is in fact optional (i.e. could be `None`). Though we could simply move the `None` check out into those methods if you prefer:
https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/cosmos/azure-cosmos/azure/cosmos/database.py#L235
https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/cosmos/azure-cosmos/azure/cosmos/cosmos_client.py#L284